src/stream/futures.rs

```diff
@@ -0,0 +1,80 @@
+use future::{Future, IntoFuture};
+use stream::Stream;
+use poll::Poll;
+use Async;
+use stack::Stack;
+use std::sync::Arc;
+use task::{self, UnparkEvent};
+
+use std::prelude::v1::*;
+
+/// An adaptor for a stream of futures to execute the futures concurrently, if
+/// possible, delivering results as they become available.
+///
+/// This adaptor will return their results in the order that they complete.
+/// This is created by the `futures` method.
+///
+#[must_use = "streams do nothing unless polled"]
+pub struct Futures<F>
+    where F: Future
+{
+    futures: Vec<F>,
+    // A list of events that will get pushed onto concurrently by our many
+    // futures. This is filled in and used with the `with_unpark_event`
+    // function.
+    stack: Arc<Stack<usize>>,
+}
+
+/// Converts a `Vec` of `IntoFuture` elements into a `Stream` over the underlying
+/// futures. The created stream will yield into values produced by those futures.
+/// Similar to `buffer_unordered` the order of the produced items will not be equal
+/// with the order of elements in the input vector.
+///
+pub fn futures<I>(futures: I) -> Futures<<I::Item as IntoFuture>::Future>
+    where I: IntoIterator,
+          I::Item: IntoFuture
+{
+    let futures = futures.into_iter().map(IntoFuture::into_future).collect::<Vec<_>>();
+    let stack = Arc::new(Stack::new());
+    for i in 0..futures.len() {
+        stack.push(i);
+    }
+    Futures {
+        futures: futures,
+        stack: stack,
+    }
+}
+
+
+impl<F> Stream for Futures<F>
+    where F: Future
+{
+    type Item = F::Item;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        if self.futures.is_empty() {
+            return Ok(Async::Ready(None));
+        }
+        let mut ready = None;
+        for (id, f) in self.futures.iter_mut().enumerate() {
+            let event = UnparkEvent::new(self.stack.clone(), id);
+            match task::with_unpark_event(event, || f.poll()) {
+                Ok(Async::NotReady) => {}
+                Ok(Async::Ready(val)) => {
+                    ready = Some((id, val));
+                    break;
+                }
+                Err(e) => {
+                    return Err(e);
+                }
+            }
+        }
+        if let Some((id, val)) = ready {
+            self.futures.swap_remove(id);
```

Comment:
It occurred to me that swap_remove is no longer needed, since the future should not be polled again.  This also means the non-allocating version which just takes a Vec could work with unpark events.